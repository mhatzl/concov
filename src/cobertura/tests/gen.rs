#[test]
fn gen_v4_read() {
    let content = include_str!("gen.xml");

    let cobertura = quick_xml::de::from_str::<crate::cobertura::schema::Coverage>(content).unwrap();

    insta::assert_ron_snapshot!(cobertura);
}

#[test]
fn gen_loose_read() {
    let content = include_str!("gen.xml");

    let cobertura =
        quick_xml::de::from_str::<crate::cobertura::schema_loose::Coverage>(content).unwrap();

    insta::assert_ron_snapshot!(cobertura);
}

#[test]
fn roundtrip() {
    let content = include_str!("gen.xml");
    let cobertura = quick_xml::de::from_str::<crate::cobertura::schema::Coverage>(content).unwrap();

    let ser_cobertura = cobertura.to_string();

    // removes whitespace formatting and possible x.0 for decimal representation
    let stripped_content = content
        .replace([' ', '\r', '\n', '\t'], "")
        .replace(".0\"", "\"");
    let stripped_cobertura = ser_cobertura
        .replace([' ', '\r', '\n', '\t'], "")
        .replace(".0\"", "\"");

    assert_eq!(
        stripped_cobertura, stripped_content,
        "Roundtrip cobertura doesn't macht original."
    );
}
