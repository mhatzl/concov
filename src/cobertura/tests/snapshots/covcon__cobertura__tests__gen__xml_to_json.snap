---
source: src/cobertura/tests/gen.rs
expression: json_content
---
{
  "line_rate": 0.85,
  "branch_rate": 0.75,
  "lines_covered": 85,
  "lines_valid": 100,
  "branches_covered": 15,
  "branches_valid": 20,
  "complexity": 1.5,
  "version": "2.1.1",
  "timestamp": "1702139945",
  "sources": {
    "source": [
      "/path/to/source1",
      "/path/to/source2"
    ]
  },
  "packages": {
    "package": [
      {
        "name": "com.example.mypackage1",
        "line_rate": 0.9,
        "branch_rate": 0.8,
        "complexity": 1.2,
        "classes": {
          "class": [
            {
              "name": "MyClass1",
              "filename": "com/example/mypackage1/MyClass1.java",
              "line_rate": 0.9,
              "branch_rate": 0.8,
              "complexity": 1.1,
              "methods": {
                "method": [
                  {
                    "name": "method1",
                    "signature": "(I)V",
                    "line_rate": 1.0,
                    "branch_rate": 1.0,
                    "complexity": 1.0,
                    "lines": {
                      "line": [
                        {
                          "number": 10,
                          "hits": 10,
                          "branch": false
                        },
                        {
                          "number": 15,
                          "hits": 5,
                          "branch": true,
                          "condition_coverage": "50%",
                          "conditions": {
                            "condition": [
                              {
                                "number": 0,
                                "type": "jump",
                                "coverage": "50%"
                              },
                              {
                                "number": 1,
                                "type": "switch",
                                "coverage": "100%"
                              }
                            ]
                          }
                        }
                      ]
                    }
                  },
                  {
                    "name": "method2",
                    "signature": "(Ljava/lang/String;)V",
                    "line_rate": 0.8,
                    "branch_rate": 0.6,
                    "complexity": 1.2,
                    "lines": {
                      "line": [
                        {
                          "number": 20,
                          "hits": 3,
                          "branch": false
                        },
                        {
                          "number": 25,
                          "hits": 0,
                          "branch": true,
                          "condition_coverage": "0%",
                          "conditions": {
                            "condition": [
                              {
                                "number": 0,
                                "type": "jump",
                                "coverage": "0%"
                              }
                            ]
                          }
                        }
                      ]
                    }
                  }
                ]
              },
              "lines": {
                "line": [
                  {
                    "number": 10,
                    "hits": 10,
                    "branch": false
                  },
                  {
                    "number": 15,
                    "hits": 5,
                    "branch": true,
                    "condition_coverage": "50%",
                    "conditions": {
                      "condition": [
                        {
                          "number": 0,
                          "type": "jump",
                          "coverage": "50%"
                        },
                        {
                          "number": 1,
                          "type": "switch",
                          "coverage": "100%"
                        }
                      ]
                    }
                  }
                ]
              }
            },
            {
              "name": "MyClass2",
              "filename": "com/example/mypackage1/MyClass2.java",
              "line_rate": 0.8,
              "branch_rate": 0.7,
              "complexity": 1.3,
              "methods": {
                "method": [
                  {
                    "name": "method1",
                    "signature": "()V",
                    "line_rate": 1.0,
                    "branch_rate": 0.8,
                    "complexity": 1.1,
                    "lines": {
                      "line": [
                        {
                          "number": 30,
                          "hits": 8,
                          "branch": false
                        },
                        {
                          "number": 35,
                          "hits": 2,
                          "branch": true,
                          "condition_coverage": "66%",
                          "conditions": {
                            "condition": [
                              {
                                "number": 0,
                                "type": "jump",
                                "coverage": "66%"
                              }
                            ]
                          }
                        }
                      ]
                    }
                  }
                ]
              },
              "lines": {
                "line": [
                  {
                    "number": 30,
                    "hits": 8,
                    "branch": false
                  },
                  {
                    "number": 35,
                    "hits": 2,
                    "branch": true,
                    "condition_coverage": "66%",
                    "conditions": {
                      "condition": [
                        {
                          "number": 0,
                          "type": "jump",
                          "coverage": "66%"
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      },
      {
        "name": "com.example.mypackage2",
        "line_rate": 0.85,
        "branch_rate": 0.75,
        "complexity": 1.4,
        "classes": {
          "class": [
            {
              "name": "AnotherClass",
              "filename": "com/example/mypackage2/AnotherClass.java",
              "line_rate": 0.85,
              "branch_rate": 0.75,
              "complexity": 1.4,
              "methods": {
                "method": [
                  {
                    "name": "anotherMethod",
                    "signature": "()I",
                    "line_rate": 1.0,
                    "branch_rate": 1.0,
                    "complexity": 1.0,
                    "lines": {
                      "line": [
                        {
                          "number": 40,
                          "hits": 10,
                          "branch": false
                        },
                        {
                          "number": 45,
                          "hits": 7,
                          "branch": true,
                          "condition_coverage": "75%",
                          "conditions": {
                            "condition": [
                              {
                                "number": 0,
                                "type": "jump",
                                "coverage": "75%"
                              }
                            ]
                          }
                        }
                      ]
                    }
                  }
                ]
              },
              "lines": {
                "line": [
                  {
                    "number": 40,
                    "hits": 10,
                    "branch": false
                  },
                  {
                    "number": 45,
                    "hits": 7,
                    "branch": true,
                    "condition_coverage": "75%",
                    "conditions": {
                      "condition": [
                        {
                          "number": 0,
                          "type": "jump",
                          "coverage": "75%"
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    ]
  }
}
